jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookieStore.java

```diff
@@ -283,64 +304,97 @@ public List<HttpCookie> match(URI uri)
             // RFC 6265, section 5.1.4, path matching algorithm.
-            if (path.equals(cookiePath))
+            if (uriPath.equals(cookiePath))
                 return true;
-            if (path.startsWith(cookiePath))
-                return cookiePath.endsWith("/") || path.charAt(cookiePath.length()) == '/';
+            if (uriPath.startsWith(cookiePath))
+                return cookiePath.endsWith("/") || uriPath.charAt(cookiePath.length()) == '/';
             return false;
         }
 
         @Override
         public boolean remove(URI uri, HttpCookie cookie)
         {
-            Key key = new Key(HttpScheme.isSecure(uri.getScheme()), uri.getHost());
+            String uriDomain = uri.getHost();
+            if (uriDomain == null)
+                return false;
+
+            String resolvedPath = resolvePath(uri, cookie);
+
+            boolean[] removed = new boolean[1];
             try (AutoLock ignored = lock.lock())
             {
-                boolean[] result = new boolean[1];
-                cookies.compute(key, (k, v) ->
+                String domain = uriDomain.toLowerCase(Locale.ENGLISH);
+                while (domain != null)
                 {
-                    if (v == null)
-                        return null;
-                    boolean removed = v.remove(cookie);
-                    result[0] = removed;
-                    return v.isEmpty() ? null : v;
-                });
-                return result[0];
+                    cookies.compute(domain, (k, v) ->
+                    {
+                        if (v == null)
+                            return null;
+
+                        Iterator<StoredHttpCookie> iterator = v.iterator();
+                        while (iterator.hasNext())
+                        {
+                            StoredHttpCookie storedCookie = iterator.next();
+                            if (uriDomain.equalsIgnoreCase(storedCookie.uri.getHost()))
+                            {
+                                if (storedCookie.path.equals(resolvedPath))
+                                {
+                                    if (storedCookie.getWrapped().getName().equals(cookie.getName()))
+                                    {
+                                        iterator.remove();
+                                        removed[0] = true;
+                                    }
+                                }
+                            }
+                        }
+
+                        return v.isEmpty() ? null : v;
+                    });
+                    domain = parentDomain(domain);
+                }
             }
+            return removed[0];
+        }
+
+        private String parentDomain(String domain)
+        {
+            int dot = domain.indexOf('.');
+            if (dot < 0)
+                return null;
+            // Remove one subdomain.
+            domain = domain.substring(dot + 1);
```

Comment:
This looks like a potential IndexOutOfBoundsException

What if the domain was `mymachine.local.` (it ends with `.`, which is 100% legal)

Reply:
The domain is only entered in the map without the final `.` (it is trimmed out) so it cannot happen.

Reply:
Hang on, if the origin server is `mymachine.local.` and the `Set-Cookie` header includes `Domain="mymachine.local."`, then you ignore the `Domain` attribute, but still store the `Cookie` against the origin server.  The rules in the RFC only state that the `Domain` attribute is ignored, presumably so that a `Set-Cookie` cannot use something like `Domain="com."` and attempt to trick the cookie store.   The map you are referring to should still have a Cookie that will only be sent back on an origin server of `mymachine.local.`.